API for MPI. First pass

AlexBuccheri opened the pull request:
Subroutines and functions added for updating species and data dependent on species ordering. Number of atoms must be kept constant. Placeholders left for support of additional functionality.

AlexBuccheri commented on prog/dftb+/api/mm/mmapi.F90:
```diff
@@ -467,6 +473,64 @@ subroutine convertAtomTypesToSpecies(typeNumbers, species, speciesNames, typeNam
 
   end subroutine convertAtomTypesToSpecies
 
+  !TODO(Alex) Consider whether this should be a member function or free function
```
Consider whether this should be a member function or free function

aradi replied:
It should remain a member function.

AlexBuccheri commented on prog/dftb+/api/mm/mmapi.F90:
```diff
@@ -467,6 +473,64 @@ subroutine convertAtomTypesToSpecies(typeNumbers, species, speciesNames, typeNam
+    !use dftbp_accuracy,    only: mc
+    use dftbp_message, only: error
+    class(TDftbPlus),               intent(inout) :: this
+    character(len=3),  allocatable, intent(in)    :: inputSpeciesNames(:)
```
Pass mc to set length?

aradi replied:
Use `character(*)`. Is the `allocatable` attribute necessary?

AlexBuccheri commented on prog/dftb+/api/mm/mmapi.F90:
```diff
@@ -467,6 +473,64 @@ subroutine convertAtomTypesToSpecies(typeNumbers, species, speciesNames, typeNam
+    
+    call this%checkInit()
+    !TODO(Alex) May not want this call here 
+    call this%checkSpeciesNames(inputSpeciesNames)
```
Check speciesNames hasn't changed here or leave to the developer to call where appropriate?

aradi replied:
Would the API handle in its current status cases, where the specie names had been changed? I think, rather not, so let's check for it explicitely.

AlexBuccheri commented on prog/dftb+/lib_dftbplus/mainapi.F90:
```diff
@@ -188,10 +191,433 @@ function nrOfAtoms()
+  !> When order of atoms changes, update arrays containing atom type indices,
+  !> and all subsequent dependencies. Updated data returned via use statements
+  subroutine updateDataDependentOnSpeciesOrdering(env, inputSpecies)
+    !TODO(Alex) These will go to the top of the module
```
Move "use module, only:" to top of current module? I like declaring per subroutine to limit scope but still appears to be declare once per .f90 file

aradi replied:
We generally place all module imports to the top level, so please do so.

AlexBuccheri commented on prog/dftb+/lib_dftbplus/mainapi.F90:
```diff
@@ -188,10 +191,433 @@ function nrOfAtoms()
+    integer, allocatable, intent(in) :: inputSpecies(:)
+    
+    !Number of atoms must be conserved    
+    @:ASSERT(size(inputSpecies) == nAtom) 
```
This could change if running in release mode: Should write function to check nAtom is conserved

aradi replied:
You can leave it as  an assertion for the moment.

AlexBuccheri commented on prog/dftb+/lib_dftbplus/mainapi.F90:
```diff
@@ -188,10 +191,433 @@ function nrOfAtoms()
+  end function updateAtomicMasses
+
+  
+  ! TODO(Alex) This contains copy and pasted code from initprogram.F90
```
Wrap code in this subroutine into smaller subroutines? Same code currently appears in initprogram.F90

aradi replied:
Yes, absolutely. One should refactor the code into a separate subroutine in `initprogram.F90` and call that. (Whenever you copy-paste, there is something wrong with the code structure :wink:)

AlexBuccheri commented on prog/dftb+/lib_dftbplus/mainapi.F90:
```diff
@@ -188,10 +191,433 @@ function nrOfAtoms()
+
+    !> TODO(Alex) Not global in initprogram. Currently hard-coded
+    !> Is check-sum for charges read externally to be used?
+    logical  :: tSkipChrgChecksum = .false.
```
Pass this in

aradi replied:
Again, we should try to refactor the program in `initprogram.F90` and call it instead of duplicating things.

AlexBuccheri commented on prog/dftb+/lib_dftbplus/mainapi.F90:
```diff
@@ -188,10 +191,433 @@ function nrOfAtoms()
+    endif
+
+    @:ASSERT(size(species0) == nAtom) 
+    !call error("Number of atoms in species array differ from current state of DFTB+")
```
Remove line

aradi replied:
Yes, please remove it.

AlexBuccheri commented on prog/dftb+/lib_dftbplus/mainapi.F90:
```diff
@@ -188,10 +191,433 @@ function nrOfAtoms()
+    !referenceN0 consistent as long as speciesName remains constant
+    call initQFromShellChrg(q0, referenceN0, species0, orb)
+
+    !TODO(Alex) This does not appear to get used in any instance in DFTB+. Remove?
```
Is qShell0 required?

aradi replied:
Probably not. Nevertheless, we must this factorize out in `initprogram.F90` to avoid code duplication.

AlexBuccheri commented on prog/dftb+/lib_dftbplus/mainapi.F90:
```diff
@@ -188,10 +191,433 @@ function nrOfAtoms()
+  end subroutine initialiseCharges
+
+  
+  ! TODO(Alex) Not sure this actually requires calling as long as the total
```
Remove TODO

aradi replied:
Yes, pls!

AlexBuccheri commented on prog/dftb+/lib_dftbplus/mainapi.F90:
```diff
@@ -188,10 +191,433 @@ function nrOfAtoms()
+  end subroutine updateBLACSDecomposition
+
+
+  ! TODO(Alex) Check if these quantities change per MD step 
```
Remove TODO

AlexBuccheri commented on prog/dftb+/lib_dftbplus/mainapi.F90:
```diff
@@ -188,10 +191,433 @@ function nrOfAtoms()
+    orb%nOrbAtom = updateAtomicOrbitals(species0)
+    call updateEquivalencyRelations(species0, iEqOrbitals, nIneqOrb, nMixElements)
+    call updateBLACSDecomposition(env, denseDesc)
+    call reallocateHSArrays(env, denseDesc, HSqrCplx, SSqrCplx, eigVecsCplx, HSqrReal, SSqrReal, eigVecsReal)
```
These quantities only require updating if number of orbitals (hence atoms) changes. No need to call in this version of the API

AlexBuccheri commented on prog/dftb+/lib_dftbplus/mainapi.F90:
```diff
@@ -188,10 +191,433 @@ function nrOfAtoms()
+  
+    if (t2Component .or. .not. tRealHS) then
+       if( (size(HSqrCplx,1) /= nLocalRows) .or. (size(HSqrCplx,2) /= nLocalCols) )then
+          write(*,*) 'Complex arrays reallocated - see mainapi.F90' 
```
Remove write statement

AlexBuccheri commented on prog/dftb+/lib_dftbplus/mainapi.F90:
```diff
@@ -188,10 +191,433 @@ function nrOfAtoms()
+       
+    else
+       if( (size(HSqrReal,1) /= nLocalRows) .or. (size(HSqrReal,2) /= nLocalCols) )then
+          write(*,*) 'Real arrays reallocated - see mainapi.F90' 
```
Remove write statement

aradi commented:
@AlexBuccheri Could you remove the superfluous `write` statements and lines? Also please note, that we work with maximal line length of 100 characters, some of your code-lines are longer. Please, fix that.

As for the refactorization in `initprogram.F90`: Do you want to give it a try?

AlexBuccheri commented:
I've finished refactoring. I'll merge Ben's changes into mine this evening (assuming he's not already addressed everything) and push

bhourahine commented:
I've not touched the initprogram.F90 changes, do you need anything done from my side there?

AlexBuccheri commented:
Thanks Ben. No, I've refactored the required code segments into subroutines. If you could take another look once I push it would be appreciated. I also have a test that can be added to the suite once I've cleaned it up.

bhourahine commented on prog/dftb+/api/mm/mmapi.F90:
```diff
@@ -21,6 +22,7 @@ module dftbp_mmapi
   use dftbp_xmlf90
   use dftbp_qdepextpotgen, only : TQDepExtPotGen, TQDepExtPotGenWrapper
   use dftbp_qdepextpotproxy, only : TQDepExtPotProxy, TQDepExtPotProxy_init
+  use dftbp_message, only: error
```
Probably a stray, as its added above

bhourahine commented on prog/dftb+/lib_dftbplus/initprogram.F90:
```diff
@@ -2310,36 +2241,7 @@ subroutine initProgramVariables(input, env)
-    if (tSccCalc) then
-      allocate(qDiffRed(nMixElements))
-      allocate(qInpRed(nMixElements))
-      allocate(qOutRed(nMixElements))
```
As its removed here it now seems not be be allocated anywhere, so causes a segfault with gnu 7.4 later on in the reduce charges routine.
 ctest -R H3_skipcheck
is one of the tests then failing.

bhourahine replied:
Ah, should be before the if (.not. tReadChrg) then block.

bhourahine replied:
Resolved in commit [71c57690](https://github.com/AlexBuccheri/dftbplus/commit/71c57690b7033e08918833e1660fe70d714820b6)

bhourahine commented:
The test fails should now be fixed

AlexBuccheri commented:
I've closed this PR as PR #420 is more up-to-date
